README.md

```diff
@@ -274,189 +285,506 @@ This is defined in libocpp/include/ocpp/v16/charge_point.hpp and takes the follo
+
+```mermaid
+sequenceDiagram
+    CSMS->>+ChargePoint : SetChargingProfileRequest(call)
+
+    ChargePoint->>+DeviceModel : SmartChargingCtrlrAvailable?
+    DeviceModel-->>-ChargePoint : Component
+
+    rect Thistle 
+    break SmartChargingCtrlrAvailable = false
+        ChargePoint-->>CSMS : Smart Charging NotSupported CallError
+    end
+    end
+
+    ChargePoint->>+SmartCharging : validate_and_add_profile(call.msg.Profile, call.msg.EVSE ID)
+
+    SmartCharging->>SmartCharging : validate_profile(Profile, EVSE ID)
+
+    rect Thistle 
+    break Invalid Profile
+        SmartCharging-->>ChargePoint : SetChargingProfileResponse: Rejected
+        ChargePoint-->>CSMS : SetChargingProfileResponse: Rejected
+    end
+    end
+
+    SmartCharging->>SmartCharging : add_profile(Profile, EVSE ID)
+
+    SmartCharging-->>-ChargePoint : SetChargingProfileResponse: Accepted
+
+    ChargePoint-->>-CSMS : SetChargingProfileResponse: Accepted
+```
+
+##### K08 Get Composite Schedule
+
+The CSMS requests the Charging Station to report the Composite Charging
+Schedule, as calculated by the Charging Station for a specific point of
+time, and may change over time due to external causes such as local
+balancing based on grid connection capacity and EVSE availablity.
+
+The Composite Schedule is the result of result of merging the time periods
+set in the `ChargingStationMaxProfile`, `ChargingStationExternalConstraints`, 
+`TxDefaultProfile` and `TxProfile` type profiles.
+
+```mermaid
+sequenceDiagram
+
+    CSMS->>+ChargePoint: GetCompositeSchedule(call)
+
+
+    ChargePoint->>+DeviceModel : ChargingScheduleChargingRateUnit?
+    DeviceModel-->>-ChargePoint : Component
+
+    rect Thistle 
+        break call.msg.chargingRateUnit is not supported
+            ChargePoint-->>CSMS : ChargingScheduleChargingRateUnitUnsupported CallError
+        end
+    end
+
+    ChargePoint->>+EvseManager : does_evse_exist(call.msg.evseId)
+    EvseManager-->>-ChargePoint : bool
+    rect Thistle 
+        break EVSE does not exist
+            ChargePoint-->>CSMS : EvseDoesNotExist CallError
+        end
+    end
+
+    ChargePoint->>+SmartChargingHandler : get_valid_profiles(call.msg.evseId)
+
+    SmartChargingHandler-->>-ChargePoint : vector<ChargingProfile>
+
+    ChargePoint->>+SmartChargingHandler : calculate_composite_schedule<br/>(vector<ChargingProfile, now, msg.duration, evseId, call.msg.chargingRateUnit)
+
+    rect Honeydew
+        loop ExternalConstraints, Max, TxDefault, and Tx Profiles
+            SmartChargingHandler->>+Profile: calculate_composite_schedule(profiles)
+            Profile-->>-SmartChargingHandler: composite_schedule 
+        end
+    end
+
+    rect Honeydew
+        loop ExternalConstraints, Max, TxDefault, and Tx Profiles
+            SmartChargingHandler->>+Profile: calculate_composite_schedule(profiles)
+            Profile-->>-SmartChargingHandler: CompositeSchedule 
+        end
+    end
```

Comment:
what is the difference between these two loops.

Reply:
Not a darn thing.